support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala

```diff
@@ -51,19 +65,32 @@ trait WebServiceHelper[Error <: Throwable] {
    */
   private def request[A](rb: Request.Builder)(implicit decoder: Decoder[A], errorDecoder: Decoder[Error], ctag: ClassTag[A]): Future[A] = {
     val req = wsPreExecute(rb).build()
-    SafeLogger.debug(s"Issuing request ${req.method} ${req.url}")
+    SafeLogger.info(s"Issuing request ${req.method} ${req.url}")
     for {
       response <- httpClient(req)
-    } yield {
-      val responseBody = response.body.string()
-      SafeLogger.debug(responseBody)
-      decode[A](responseBody) match {
-        case Left(err) => throw decodeError(responseBody).right.getOrElse(
-          WebServiceHelperError[A](response.code(), responseBody)
-        )
-        case Right(value) => value
+      decodedResponse <- {
+        val code = response.code().toString
+        val failableDecodedResponse = for {
+          body <- Option(response.body).toRight(WebServiceHelperError(code, "", "no body")).toTry
+          contentType <- Option(body.contentType()).toRight(WebServiceHelperError(code, "", "no content type")).toTry
+          bodyString <- Try(body.string())
+          _ = SafeLogger.info(s"response $code body: ${bodyString.length} bytes")
+          _ <- if (contentType.`type`() == "application" && contentType.subtype() == "json") Success(())
+          else Failure(WebServiceHelperError(code, bodyString, s"wrong content type"))
+          result <- code.splitAt(1) match {
+            case ("2", _) => decode[A](bodyString).left.map { err =>
```

Comment:
Could have left `code` as an `int` and then we don't have to split a string and match on that.  There's also an `isSuccessful` method by the looks of it, but no equivalent for client errors.

Reply:
yeah isSuccessful seems to be the only one in okhttp, unfortunately no nice ADT of response codes.
I agree I could have left as a int, but I'd still have to do `code % 100` and check it against 2 and 4, but I don't entirely feel like arithmetic makes sense on a response code.  More concretely it opens us up to off by one errors.  In a string each character is opaque when splitting so if it works in one case it should work in them all.
that's my reasoning but happy to use mod instead if you feel strongly!

Reply:
Maybe it would be easier to read quickly as an `if...else` with `code >= 200 && code < 300`? Separating out as Tom suggested is probably more meaningful though.

Reply:
```scala
      result <-
        if (code >= 200 && code < 300) 
          decode[A](bodyString).left.map { err =>
            WebServiceHelperError[A](code, bodyString, s"failed to parse response: $err")
          }.toTry
        else if (code >= 400 && code < 500) 
  Failure((decodeError(bodyString).right.toOption).getOrElse(
            WebServiceClientError(code, bodyString))
          )
        else Failure(WebServiceHelperError[A](code, bodyString, "unrecognised response code"))
      }
```

Reply:
see what you mean about the WTF factor, I have an idea to make it read better

Reply:
```scala
      result <- code(0) + "xx" match {
        case "2xx" => decode[A](bodyString).left.map { err =>
          WebServiceHelperError[A](code, bodyString, s"failed to parse response: $err")
        }.toTry
        case "4xx" => Failure((decodeError(bodyString).right.toOption).getOrElse(
          WebServiceClientError(code, bodyString))
        )
        case _ => Failure(WebServiceHelperError[A](code, bodyString, "unrecognised response code"))
      }
```